### PR TITLE
Feature: endpoint to manage SSO users

### PR DESCRIPTION
**Pull request by ShellyXueHan**
Has not yet included documentation or unit test for the feature, because they are temporary. Once settle down with the way we want the sso endpoint to behave, will wrap up with doc and test cases.

Deliverables:
1. Endpoints as `/api/v1/sso`:
  1.1 get relevant SSO user info by email
  1.2 get relevant SSO user info by user id
  1.3 update SSO user profiles -> email, first and last name (mandatory)
2. sso utility functions for users and groups
3. clean up env vars

/bot-ignore-length

### CODE REVIEW

File: src/config/index.js [hunk 1]
```diff
@@ -44,8 +44,10 @@ nconf.overrides({
   environment: env,
   host: process.env.HOST || '127.0.0.1',
   port: process.env.PORT || defaultPort,
-  sso: {
-    clientSecret: process.env.SSO_CLIENT_SECRET,
+  ssoSA: {
```
Comment (patricksimonian) [inline src/config/index.js:47]: sso Service Agent?

File: src/constants.js [hunk 2]
```diff
@@ -25,4 +25,19 @@ const ENVIRONMENTS = {
...
+};
+
+const SSO_REQUEST = {
+  CONTENT_TYPE_FORM: 'application/x-www-form-urlencoded',
+  CONTENT_TYPE_JSON: 'application/json',
+  GRANT_TYPE: 'client_credentials',
+  CLIENT_ID: 'admin-cli',
+};
+
+module.exports = { ENVIRONMENTS, SSO_SUB_URI, SSO_REQUEST };
```
Comment (patricksimonian) [inline src/constants.js]: mixing common js `module.exports` here with `export` syntax in another file `src/libs/sso-utils.js`

File: src/libs/sso-utils.js [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+    throw new Error('A URL must be provided');
+  }
+  if (!(credentials.token || (credentials.username && credentials.password))) {
+    throw new Error('An auth info must be provided, such as username-password pair or token.');
+  }
+};
+
+export const checkUserProfile = (userInfo, userGroups = [], userIdps = []) => {
+  let valid = false;
+  try {
```
Comment (patricksimonian) [inline src/libs/sso-utils.js]: why try catch? doesn't seem like anything will throw here

File: src/router/routes/ssoUsers.js [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+import config from '../../config';
+import { SSO_SUB_URI } from '../../constants';
+import { getSAToken, getUserInfoByEmail, getUserInfoById, updateUser } from '../../libs/sso-utils';
+
+const router = new Router();
+
+// Either get SSO user by email or by ID:
+router.get(
+  '/user',
+  asyncMiddleware(async (req, res) => {
```
Comment (patricksimonian) [inline src/router/routes/ssoUsers.js:36]: are there any differences between the two `asyncMiddleware` they look pretty similar.
I believe you can register the middleware by using `router.use(asyncMiddleware)`

Comment (patricksimonian) [reply src/router/routes/ssoUsers.js:36]: oops never mind

File: src/router/routes/ssoUsers.js [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    const { email } = req.query;
+
+    if (!email) {
+      throw errorWithCode('Please provide the email of the SSO user you are looking for.', 400);
+    }
+
+    logger.info(`Looking of user of ${email}`);
+    const SACredentials = config.get('ssoSA');
```
Comment (patricksimonian) [inline src/router/routes/ssoUsers.js]: constant `'ssoSa'`

**Review by patricksimonian — CHANGES_REQUESTED** [review]
Looks good.  Only change I'd recommend is to change over the common js in the the constants file to the es6 syntax since everywhere else its being used.

So I'd put an issue in for refactoring tasks later on. Perhaps moving the request callback functions in `src/router/routes/ssoUsers.js` into a seperate place so that they are testable later on.

File: src/libs/sso-utils.js (hunk 3 above)
Comment (ShellyXueHan) [reply src/libs/sso-utils.js]: right, updating that

File: src/router/routes/ssoUsers.js (hunk 5 above)
Comment (ShellyXueHan) [reply src/router/routes/ssoUsers.js]: haha you got me

File: src/config/index.js (hunk 1 above)
Comment (ShellyXueHan) [reply src/config/index.js:47]: yes

**Review by patricksimonian — APPROVED** [review]
looks rad!

**Pull request merged**